src/core/ext/transport/binder/wire_format/wire_writer.cc

```diff
@@ -66,119 +80,301 @@ absl::Status WireWriterImpl::WriteTrailingMetadata(const Transaction& tx,
+  if (is_transacting_) {
+    // This can happen because NDK might call our registered callback function
+    // in the same thread while we are telling it to send a transaction
+    // `is_transacting_` will be true. `Binder::Transact` is now being called on
+    // the same thread or the other thread. We are currently in the call stack
+    // of other transaction, Liveness of ACK is still guaranteed even if this is
+    // a race with another thread.
+    gpr_log(
+        GPR_INFO,
+        "Scheduling ACK transaction instead of directly execute it to avoid "
+        "deadlock.");
+    auto args = new RunScheduledTxArgs();
+    args->writer = this;
+    args->tx = RunScheduledTxArgs::AckTx();
+    absl::get<RunScheduledTxArgs::AckTx>(args->tx).num_bytes = num_bytes;
+    auto cl = GRPC_CLOSURE_CREATE(RunScheduledTx, args, nullptr);
+    combiner_->Run(cl, GRPC_ERROR_NONE);
+    return absl::OkStatus();
+  }
+  // Otherwise, we can directly send ack.
+  absl::Status result = MakeBinderTransaction(
+      BinderTransportTxCode(BinderTransportTxCode::ACKNOWLEDGE_BYTES),
+      [num_bytes](WritableParcel* parcel) {
+        RETURN_IF_ERROR(parcel->WriteInt64(num_bytes));
+        return absl::OkStatus();
+      });
+  if (!result.ok()) {
+    gpr_log(GPR_ERROR, "Failed to make binder transaction %s",
+            result.ToString().c_str());
+  }
+  return result;
 }
 
 void WireWriterImpl::OnAckReceived(int64_t num_bytes) {
-  grpc_core::MutexLock lock(&mu_);
-  num_acknowledged_bytes_ = std::max(num_acknowledged_bytes_, num_bytes);
-  cv_.Signal();
+  // Ensure combiner will be run if this is not called from top-level gRPC API
+  // entrypoint.
+  grpc_core::ExecCtx exec_ctx;
+  gpr_log(GPR_INFO, "OnAckReceived %ld", num_bytes);
+  // Do not try to obtain `write_mu_` in this function. NDKBinder might invoke
+  // the callback to notify us about new incoming binder transaction when we are
+  // sending transaction. i.e. `write_mu_` might have already been acquired by
+  // this thread.
+  {
+    grpc_core::MutexLock lock(&flow_control_mu_);
+    num_acknowledged_bytes_ = std::max(num_acknowledged_bytes_, num_bytes);
+    int64_t num_outgoing_bytes = num_outgoing_bytes_;
+    if (num_acknowledged_bytes_ > num_outgoing_bytes) {
+      gpr_log(GPR_ERROR,
+              "The other end of transport acked more bytes than we ever sent, "
+              "%ld > %ld",
+              num_acknowledged_bytes_, num_outgoing_bytes);
+    }
+  }
+  TryScheduleTransaction();
+}
+
+void WireWriterImpl::TryScheduleTransaction() {
+  while (true) {
+    grpc_core::MutexLock lock(&flow_control_mu_);
+    // An estimation of number of bytes of non-acked traffic we will eventually
+    // have, assuming all tasks in combiner will be executed and we receive no
+    // new ACK from the other end of transport.
+    int64_t num_non_acked_bytes_possible =
+        (num_outgoing_bytes_ + num_non_acked_tx_in_combiner_ * kBlockSize) -
```

Comment:
Explain how you came up with the estimation formula in the previous comment.

Reply:
Done. I split it into 3 variables and add comments to them